### PR TITLE
Auto case update rules compatible with SQL

### PR DESCRIPTION
**Pull request by snopoke**
@sravfeyn 

@gcapalbo 

cc @dimagi/scale-team

### CODE REVIEW

File: corehq/form_processor/models.py [hunk 1]
```diff
@@ -536,8 +536,9 @@ def dynamic_case_properties(self):
 
     def to_json(self):
         from .serializers import CommCareCaseSQLSerializer
-        serializer = CommCareCaseSQLSerializer(self)
-        return serializer.data
+        data = CommCareCaseSQLSerializer(self).data
+        data.update(self.case_json)
+        return data
```
Comment (snopoke) [inline corehq/form_processor/models.py]: @czue FYI

**Comment by snopoke** [comment]
@gcapalbo I notice that the check functions in [AutomaticUpdateRuleCriteria](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/data_interfaces/models.py#L123) use two different ways of getting case properties:
* case.to_json().get(self.property_name)
* case.get_case_property(self.property_name)

Is there a specific reason for that? Also, are these properties always going to be dynamic case properties are could they be case meta properties?

File: corehq/form_processor/models.py (hunk 1 above)
Comment (czue) [inline corehq/form_processor/models.py]: cool that's more compact than [my version](https://github.com/dimagi/commcare-hq/commit/17e339b0662e769375d1af31a096d6ecbb82869c). I was worried about clobbering the global namespace

Comment (snopoke) [inline corehq/form_processor/models.py]: :+1: yours is better

**Comment by gcapalbo** [comment]
@snopoke they can all use case.to_json(). i had to update two of them to use case.to_json() because jsonobject was automatically casting certain types and causing matching issues, but i didn't update the other two. fine to update the other two to make them consistent.

the properties referenced can be dynamic properties as well as static ones (like name, owner_id, etc.)

**Comment by gcapalbo** [comment]
@snopoke just a few comments otherwise looks good!  is there anything special that needs to be done to run the sql tests or does it work automatically (locally and with travis)?

**Comment by snopoke** [comment]
Nothing special to make the tests run. Each test gets run twice, once for
couch forms/cases and once for SQL

File: corehq/apps/data_interfaces/tests/test_auto_case_updates.py [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
+from casexml.apps.case.signals import case_post_save
 from corehq.apps.data_interfaces.models import (AutomaticUpdateRule,
     AutomaticUpdateRuleCriteria, AutomaticUpdateAction)
 from corehq.apps.data_interfaces.tasks import run_case_update_rules_for_domain
 from datetime import datetime, date
-from dimagi.utils.parsing import json_format_datetime
+
+from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.form_processor.tests.utils import run_with_all_backends, FormProcessorTestUtils, set_case_property_directly
```
Comment (dimagimon) [inline corehq/apps/data_interfaces/tests/test_auto_case_updates.py]: E501 line too long (119 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by NoahCarnahan** [comment]
Killed your build to get travis to run the P1 build more quickly. Sorry :(

**Comment by gcapalbo** [comment]
(restarted build)
@snopoke this is good from my side - was there anything else you wanted to add or are we good to merge?

**Comment by gcapalbo** [comment]
actually there is one more thing - i don't see the update you made anymore to to_json(): https://github.com/dimagi/commcare-hq/blob/auto-case/corehq/form_processor/models.py#L268

can you confirm that to_json returns dynamic properties + native ones?

**Comment by snopoke** [comment]
@gcapalbo GTG - the `to_json` stuff is reverted to use what's on master: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/models.py#L537

**Comment by gcapalbo** [comment]
great, i'm ready to merge but looks like build failed

**Comment by gcapalbo** [comment]
build failure might be related to https://github.com/dimagi/commcare-hq/pull/10593#issuecomment-191832561

**Comment by snopoke** [comment]
rebased off master

**Comment by gcapalbo** [comment]
(merged master to fix the build, once build passes will merge)

**Comment by gcapalbo** [comment]
(kicked build again)

**Comment by snopoke** [comment]
:white_check_mark: finally

**Comment by snopoke** [comment]
self merging

**Comment by czue** [comment]
:tada:

**Comment by gcapalbo** [comment]
:+1:

**Pull request merged**